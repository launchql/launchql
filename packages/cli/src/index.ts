#!/usr/bin/env node
import { CLI, CLIOptions } from 'inquirerer';

import { commands } from './commands';

export const options: Partial<CLIOptions> = {
  minimistOpts: {
    alias: {
      v: 'version',
      h: 'help'
    }
  }
};

const app = new CLI(commands, options);

app.run().then(()=> {
  // all done!
}).catch(error => {
  // Should not reach here with the new CLI error handling pattern
  // But keep as fallback for unexpected errors
  console.error('Unexpected error:', error);
  process.exit(1);
});