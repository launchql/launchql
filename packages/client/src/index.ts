import { Pool, PoolClient } from 'pg';

import env from './env';

export class Database {
  private static instance: Database;
  private pool: Pool;

  constructor() {
    if (Database.instance) {
      return Database.instance;
    }

    const pgPoolConfig = {
      connectionString: env.DATABASE_URL,
    };
    this.pool = new Pool(pgPoolConfig);

    // Ensure the pool is closed on process termination
    process.on('SIGTERM', async () => {
      await this.shutdown();
    });

    Database.instance = this;
    return this;
  }

  /**
   * Executes a callback function within a database transaction.
   * @param fn - A callback function that receives a PoolClient to perform database operations.
   */
  async withTransaction(fn: (client: PoolClient) => Promise<void>): Promise<void> {
    const client = await this.pool.connect();
    try {
      await client.query('BEGIN');
      try {
        await fn(client);
        await client.query('COMMIT');
      } catch (e) {
        console.error('Error during transaction:', e);
        await client.query('ROLLBACK');
        throw e; // Re-throw the error to propagate it
      }
    } finally {
      client.release();
    }
  }

  /**
   * Shuts down the connection pool.
   */
  async shutdown(): Promise<void> {
    await this.pool.end();
  }
}