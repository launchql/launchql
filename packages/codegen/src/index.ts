import { generateCode } from './codegen';
import getIntrospectionRows, { GetIntrospectionRowsOptions } from './introspect';
import { DatabaseObject } from './types';

(async () => {
  const options: GetIntrospectionRowsOptions = {
    connectionString: 'postgresql://postgres:password@localhost:5432/mydb',
    introspectionOptions: {
      pgLegacyFunctionsOnly: false,
      pgIgnoreRBAC: true,
    },
    namespacesToIntrospect: ['collections_public'],
    includeExtensions: false,
  };

  try {
    // Fetch introspection rows
    const rows: DatabaseObject[] = await getIntrospectionRows(options);
    console.log('Introspection Rows Fetched:', rows);

    // Generate TypeScript code
    const codegenOptions = {
      includeTimestamps: true,
      includeUUID: true,
    };
    const generatedCode = generateCode(rows, codegenOptions);

    console.log('Generated TypeScript Code:');
    console.log(generatedCode);
  } catch (error) {
    console.error('Failed to fetch introspection rows or generate code:', error);
  }
})();
