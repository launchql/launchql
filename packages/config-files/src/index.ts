export * from './types';
export * from './plan';
export * from './config';
export * from './sql';
export * from './extension';
export * from './sql-scripts';