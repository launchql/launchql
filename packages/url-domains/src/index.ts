export * from './requests';
export * from './urls';
export * from './middleware';