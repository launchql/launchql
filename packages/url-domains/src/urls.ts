const ipRegExp = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/;

export const parseUrl = (input: string | URL) => {
  const url = typeof input === 'string' ? new URL(input) : input;
  let hostname = url.hostname.replace(/^www\./, '');

  const parts = hostname.split('.');
  let domain: string | null = null;
  let subdomains: string[] | null = null;

  if (hostname === 'localhost') {
    domain = 'localhost';
    subdomains = [];
  } else if (hostname.endsWith('.localhost')) {
    domain = 'localhost';
    subdomains = parts.slice(0, -1); // everything before 'localhost'
  } else if (ipRegExp.test(hostname)) {
    domain = hostname;
    subdomains = [];
  } else if (parts.length >= 2) {
    domain = parts.slice(-2).join('.');
    subdomains = parts.slice(0, -2);
  } else {
    domain = hostname;
    subdomains = [];
  }

  return {
    hostname: url.hostname,
    domain,
    subdomains
  };
};
