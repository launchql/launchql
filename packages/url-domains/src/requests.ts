import { Request } from 'express';

import { parseUrl } from './urls';

export const parseReq = (req: Request) => {
  const fullUrl = `${req.protocol}://${req.get('host')}${req.originalUrl}`;
  return parseUrl(fullUrl);
};
