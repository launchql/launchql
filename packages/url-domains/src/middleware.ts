import { Request, Response, NextFunction } from 'express';
import { parseReq } from './requests';

interface UrlDomains {
  [key: string]: string | string[];
}

declare global {
  namespace Express {
    interface Request {
      urlDomains?: UrlDomains;
    }
  }
}

export const middleware = ({
  hostname = 'hostname',
  subdomains = 'subdomains',
  domain = 'domain'
} = {}) => {
  return async (req: Request, res: Response, next: NextFunction) => {
    const parsed = parseReq(req);
    console.log(parsed);
    req.urlDomains = {
      [hostname]: parsed.hostname,
      [subdomains]: parsed.subdomains,
      [domain]: parsed.domain
    };
    next();
  };
};
