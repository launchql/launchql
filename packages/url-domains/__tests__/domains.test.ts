import { parseReq, middleware as parseMiddleware } from '../src';
import cases from 'jest-in-case';

const fixtures = [
  {
    protocol: 'http',
    host: 'www.admin.localhost:8888',
    originalUrl: '/'
  },
  { protocol: 'http', host: 'admin.localhost:8888', originalUrl: '/' },
  { protocol: 'http', host: 'localhost:8888', originalUrl: '/' },
  {
    protocol: 'http',
    host: 'admin.basecrypt.com:8888',
    originalUrl: '/'
  },
  {
    protocol: 'http',
    host: 'admin.web.basecrypt.com:8888',
    originalUrl: '/'
  },
  {
    protocol: 'http',
    host: '127.0.0.1:8888',
    originalUrl: '/graphiql?yo=1'
  },
  {
    protocol: 'http',
    host: 'coolio.webby.localhost:8888',
    originalUrl: '/graphiql?yo=1'
  }
];

cases(
  'parses URLs',
  options => {
    const req = {
      ...options,
      get: () => {
        return options.host;
      }
    };
    // @ts-ignore
    expect(parseReq(req)).toMatchSnapshot();
  },
  fixtures.map(fix => {
    return { ...fix, name: fix.host };
  })
);

cases(
  'parses reqs',
  async options => {
    const req = {
      ...options,
      get: () => {
        return options.host;
      }
    };
    // @ts-ignore
    await parseMiddleware()(req, null, () => {});
    expect(req).toMatchSnapshot();
  },
  fixtures.map(fix => {
    return { ...fix, name: fix.host };
  })
);