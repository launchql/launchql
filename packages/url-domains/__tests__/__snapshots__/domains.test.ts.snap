// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses URLs 127.0.0.1:8888 1`] = `
{
  "domain": "127.0.0.1",
  "hostname": "127.0.0.1",
  "subdomains": [],
}
`;

exports[`parses URLs admin.basecrypt.com:8888 1`] = `
{
  "domain": "basecrypt.com",
  "hostname": "admin.basecrypt.com",
  "subdomains": [
    "admin",
  ],
}
`;

exports[`parses URLs admin.localhost:8888 1`] = `
{
  "domain": "admin.localhost",
  "hostname": "admin.localhost",
  "subdomains": [],
}
`;

exports[`parses URLs admin.web.basecrypt.com:8888 1`] = `
{
  "domain": "basecrypt.com",
  "hostname": "admin.web.basecrypt.com",
  "subdomains": [
    "admin",
    "web",
  ],
}
`;

exports[`parses URLs coolio.webby.localhost:8888 1`] = `
{
  "domain": "webby.localhost",
  "hostname": "coolio.webby.localhost",
  "subdomains": [
    "coolio",
  ],
}
`;

exports[`parses URLs localhost:8888 1`] = `
{
  "domain": "localhost",
  "hostname": "localhost",
  "subdomains": [],
}
`;

exports[`parses URLs www.admin.localhost:8888 1`] = `
{
  "domain": "admin.localhost",
  "hostname": "www.admin.localhost",
  "subdomains": [],
}
`;

exports[`parses reqs 127.0.0.1:8888 1`] = `
{
  "get": [Function],
  "host": "127.0.0.1:8888",
  "name": "127.0.0.1:8888",
  "originalUrl": "/graphiql?yo=1",
  "protocol": "http",
  "urlDomains": {
    "domain": "127.0.0.1",
    "hostname": "127.0.0.1",
    "subdomains": [],
  },
}
`;

exports[`parses reqs admin.basecrypt.com:8888 1`] = `
{
  "get": [Function],
  "host": "admin.basecrypt.com:8888",
  "name": "admin.basecrypt.com:8888",
  "originalUrl": "/",
  "protocol": "http",
  "urlDomains": {
    "domain": "basecrypt.com",
    "hostname": "admin.basecrypt.com",
    "subdomains": [
      "admin",
    ],
  },
}
`;

exports[`parses reqs admin.localhost:8888 1`] = `
{
  "get": [Function],
  "host": "admin.localhost:8888",
  "name": "admin.localhost:8888",
  "originalUrl": "/",
  "protocol": "http",
  "urlDomains": {
    "domain": "admin.localhost",
    "hostname": "admin.localhost",
    "subdomains": [],
  },
}
`;

exports[`parses reqs admin.web.basecrypt.com:8888 1`] = `
{
  "get": [Function],
  "host": "admin.web.basecrypt.com:8888",
  "name": "admin.web.basecrypt.com:8888",
  "originalUrl": "/",
  "protocol": "http",
  "urlDomains": {
    "domain": "basecrypt.com",
    "hostname": "admin.web.basecrypt.com",
    "subdomains": [
      "admin",
      "web",
    ],
  },
}
`;

exports[`parses reqs coolio.webby.localhost:8888 1`] = `
{
  "get": [Function],
  "host": "coolio.webby.localhost:8888",
  "name": "coolio.webby.localhost:8888",
  "originalUrl": "/graphiql?yo=1",
  "protocol": "http",
  "urlDomains": {
    "domain": "webby.localhost",
    "hostname": "coolio.webby.localhost",
    "subdomains": [
      "coolio",
    ],
  },
}
`;

exports[`parses reqs localhost:8888 1`] = `
{
  "get": [Function],
  "host": "localhost:8888",
  "name": "localhost:8888",
  "originalUrl": "/",
  "protocol": "http",
  "urlDomains": {
    "domain": "localhost",
    "hostname": "localhost",
    "subdomains": [],
  },
}
`;

exports[`parses reqs www.admin.localhost:8888 1`] = `
{
  "get": [Function],
  "host": "www.admin.localhost:8888",
  "name": "www.admin.localhost:8888",
  "originalUrl": "/",
  "protocol": "http",
  "urlDomains": {
    "domain": "admin.localhost",
    "hostname": "www.admin.localhost",
    "subdomains": [],
  },
}
`;
